server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -419,34 +432,77 @@ public ClusterState getLastAcceptedState() {
+            }
+            return writer;
+        }
+
         @Override
         public void setCurrentTerm(long currentTerm) {
-            persistenceWriter.commit(currentTerm, lastAcceptedState.version());
+            reloadWriterIfNecessary();
+            try {
+                if (writeNextStateFully) {
+                    getWriterSafe().writeFullStateAndCommit(currentTerm, lastAcceptedState);
+                    writeNextStateFully = false;
+                } else {
+                    getWriterSafe().commit(currentTerm, lastAcceptedState.version());
+                }
+            } catch (Exception e) {
+                handleExceptionOnWrite(e);
+            }
             this.currentTerm = currentTerm;
         }
 
         @Override
         public void setLastAcceptedState(ClusterState clusterState) {
+            reloadWriterIfNecessary();
             try {
-                if (clusterState.term() != lastAcceptedState.term()) {
-                    assert clusterState.term() > lastAcceptedState.term() : clusterState.term() + " vs " + lastAcceptedState.term();
-                    // In a new currentTerm, we cannot compare the persisted metadata's lastAcceptedVersion to those in the new state, so
-                    // it's simplest to write everything again.
-                    persistenceWriter.writeFullStateAndCommit(currentTerm, clusterState);
+                if (writeNextStateFully) {
+                    getWriterSafe().writeFullStateAndCommit(currentTerm, clusterState);
+                    writeNextStateFully = false;
                 } else {
-                    // Within the same currentTerm, we _can_ use metadata versions to skip unnecessary writing.
-                    persistenceWriter.writeIncrementalStateAndCommit(currentTerm, lastAcceptedState, clusterState);
+                    if (clusterState.term() != lastAcceptedState.term()) {
+                        assert clusterState.term() > lastAcceptedState.term() : clusterState.term() + " vs " + lastAcceptedState.term();
+                        // In a new currentTerm, we cannot compare the persisted metadata's lastAcceptedVersion to those in the new state,
+                        // so it's simplest to write everything again.
+                        getWriterSafe().writeFullStateAndCommit(currentTerm, clusterState);
+                    } else {
+                        // Within the same currentTerm, we _can_ use metadata versions to skip unnecessary writing.
+                        getWriterSafe().writeIncrementalStateAndCommit(currentTerm, lastAcceptedState, clusterState);
+                    }
                 }
-            } catch (IOException e) {
-                throw new UncheckedIOException(e);
+            } catch (Exception e) {
+                handleExceptionOnWrite(e);
             }
 
             lastAcceptedState = clusterState;
         }
 
+        private void reloadWriterIfNecessary() {
```

Comment:
I suggest combining this with `getWriterSafe()`, because in all cases we call `reloadWriterIfNecessary()` followed (essentially) immediately by `getWriterSafe()`. Why not return the writer from this method instead?

Reply:
good idea, done in 333f42a4178